### PR TITLE
Better Vocabulary Support

### PR DESCRIPTION
**Pull request by robbielamb-wf**
## Ultimate problem:
Using custom meta-schemas and vocabularies are difficult.

## How it was fixed:
Refactor some code to pave the way for custom vocabularies.

## Potential areas of regression:

---

> __FYA:__ @michaelcarter-wf

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/json_schema/json_schema.dart [hunk 1]
```diff
@@ -1018,62 +1018,141 @@ class JsonSchema {
...
-      r'$defs': (JsonSchema s, dynamic v) => s._setDefs(v),
+      r'$ref': (JsonSchema s, dynamic v) => s._setRef(v),
       r'$recursiveRef': (JsonSchema s, dynamic v) => s._setRecursiveRef(v),
       r'$recursiveAnchor': (JsonSchema s, dynamic v) => s._setRecursiveAnchor(v),
       r'$vocabulary': (JsonSchema s, dynamic v) => s._setVocabulary(v),
-
-      // Added or changed in draft2019_09: Applicator Vocabulary
-      'dependentSchemas': (JsonSchema s, dynamic v) => s._setDependentSchemas(v),
+      r'$comment': (JsonSchema s, dynamic v) => s._setComment(v),
+      r'$defs': (JsonSchema s, dynamic v) => s._setDefs(v),
```
Comment (robbielamb-wf) [inline lib/src/json_schema/json_schema.dart:1040]: These items are ordered the same way as the corresponding vocab file.

File: lib/src/json_schema/json_schema.dart [hunk 2]
```diff
@@ -1018,62 +1018,141 @@ class JsonSchema {
...
+  static Map<String, SchemaPropertySetter> _draft2019_combined = Map()
+    ..addAll(_draft2019_core)
+    ..addAll(_draft2019_applicator)
+    ..addAll(_draft2019_content)
+    ..addAll(_draft2019_format)
+    ..addAll(_draft2019_metadata)
+    ..addAll(_draft2019_validation);
 
-      // Added or changed in draft2019_09: Core Vocabulary
+  static Map<String, SchemaPropertySetter> _draft2019_core = Map<String, SchemaPropertySetter>()
```
Comment (tomdeering-wf) [inline lib/src/json_schema/json_schema.dart:1030]: It's actually really nice to have a concise synopsis of what is in each vocabulary!

**Review by tomdeering-wf — COMMENTED** [review]
Me likey

**Review by michaelcarter-wf — COMMENTED** [review]
looks awesome!

File: lib/src/json_schema/type_validators.dart [hunk 3]
```diff
@@ -84,13 +84,14 @@ class TypeValidators {
...
-  static SchemaVersion jsonSchemaVersion4Or6Or7(String key, dynamic value) {
+  static SchemaVersion builtInSchemaVersion(String key, dynamic value) {
     string(key, value);
     final schemaVersion = SchemaVersion.fromString(value);
     if (schemaVersion != null) {
       return schemaVersion;
     }
-    throw FormatExceptions.error('Only draft 4, draft 6, and draft 7 schemas supported');
+    throw FormatExceptions.error(
+        'Only draft 4, draft 6, draft 7, draft 2019, draft 2020, and custom schemas supported');
```
Comment (michaelcarter-wf) [inline lib/src/json_schema/type_validators.dart:94]:
```suggestion
        'Only draft 4, draft 6, draft 7, draft 2019-09, draft 2020-12, and custom schemas supported');
```

File: lib/src/json_schema/json_schema.dart [hunk 4]
```diff
@@ -76,24 +76,34 @@ class JsonSchema {
...
     );
   }
 
   JsonSchema._fromRootBool(this._schemaBool, SchemaVersion schemaVersion,
-      {Uri fetchedFromUri, bool isSync = false, Map<String, JsonSchema> refMap, RefProvider refProvider}) {
+      {Uri fetchedFromUri,
+      bool isSync = false,
+      Map<String, JsonSchema> refMap,
+      RefProvider refProvider,
+      Map<Uri, bool> metaschemaVocabulayr}) {
```
Comment (michaelcarter-wf) [inline lib/src/json_schema/json_schema.dart:99]:
```suggestion
      Map<Uri, bool> metaschemaVocabulay}) {
```

File: lib/src/json_schema/json_schema.dart [hunk 5]
```diff
@@ -76,24 +76,34 @@ class JsonSchema {
   }
 
   JsonSchema._fromRootMap(this._schemaMap, SchemaVersion schemaVersion,
-      {Uri fetchedFromUri, bool isSync = false, Map<String, JsonSchema> refMap, RefProvider refProvider}) {
+      {Uri fetchedFromUri,
+      bool isSync = false,
+      Map<String, JsonSchema> refMap,
+      RefProvider refProvider,
+      Map<Uri, bool> metaschemaVocabulary}) {
```
Comment (michaelcarter-wf) [inline lib/src/json_schema/json_schema.dart:83]: Comma here will help the formatter make this look better:
```suggestion
      Map<Uri, bool> metaschemaVocabulary,}) {
```

File: lib/src/json_schema/json_schema.dart [hunk 6]
```diff
@@ -76,24 +76,34 @@ class JsonSchema {
...
     );
   }
 
   JsonSchema._fromRootBool(this._schemaBool, SchemaVersion schemaVersion,
-      {Uri fetchedFromUri, bool isSync = false, Map<String, JsonSchema> refMap, RefProvider refProvider}) {
+      {Uri fetchedFromUri,
+      bool isSync = false,
+      Map<String, JsonSchema> refMap,
+      RefProvider refProvider,
+      Map<Uri, bool> metaschemaVocabulary}) {
```
Comment (michaelcarter-wf) [inline lib/src/json_schema/json_schema.dart:99]: and here
```suggestion
      Map<Uri, bool> metaschemaVocabulary,}) {
```

File: lib/src/json_schema/json_schema.dart [hunk 7]
```diff
@@ -227,40 +240,44 @@ class JsonSchema {
       _path = '#';
       final String refString = '${_uri ?? ''}$_path';
       _addSchemaToRefMap(refString, this);
-      _thisCompleter = Completer<JsonSchema>();
+
+      schemaString = _schemaMap.containsKey(r'$schema') ? _schemaMap[r'$schema'] : null;
+      _resolveMetaSchemasForVocabulary(schemaString, _schemaVersion); // For Vocabulary
```
Comment (michaelcarter-wf) [inline lib/src/json_schema/json_schema.dart:245]:
```suggestion
      _resolveMetaSchemasForVocabulary(schemaString, _schemaVersion);
```

**Review by michaelcarter-wf — APPROVED** [review]
+1, nits

File: lib/src/json_schema/json_schema.dart [hunk 8]
```diff
@@ -1018,62 +1085,167 @@ class JsonSchema {
...
+    ..addAll({
+      "https://json-schema.org/draft/2019-09/vocab/core": _draft2019_core,
+      "https://json-schema.org/draft/2019-09/vocab/applicator": _draft2019_applicator,
+      "https://json-schema.org/draft/2019-09/vocab/validation": _draft2019_validation,
+      "https://json-schema.org/draft/2019-09/vocab/meta-data": _draft2019_metadata,
+      "https://json-schema.org/draft/2019-09/vocab/format": _draft2019_format,
+      "https://json-schema.org/draft/2019-09/vocab/content": _draft2019_content
+    });
 
+  // TODO: Think about doing something with the deprecated keywords in this map.
```
Comment (tomdeering-wf) [inline lib/src/json_schema/json_schema.dart:1185]: Is this something we can safely postpone? Just checking.

File: lib/src/json_schema/type_validators.dart [hunk 9]
```diff
@@ -84,13 +84,14 @@ class TypeValidators {
...
-  static SchemaVersion jsonSchemaVersion4Or6Or7(String key, dynamic value) {
+  static SchemaVersion builtInSchemaVersion(String key, dynamic value) {
     string(key, value);
     final schemaVersion = SchemaVersion.fromString(value);
     if (schemaVersion != null) {
       return schemaVersion;
     }
-    throw FormatExceptions.error('Only draft 4, draft 6, and draft 7 schemas supported');
+    throw FormatExceptions.error(
+        'Only draft 4, draft 6, draft 7, draft 2019-09, draft 2020-12, and custom schemas supported');
```
Comment (tomdeering-wf) [inline lib/src/json_schema/type_validators.dart:94]: `custom schemas`, eh? Did you mean vocabularies?

Comment (robbielamb-wf) [reply lib/src/json_schema/type_validators.dart:94]: I did mean schemas here. A custom schema is required to use a custom vocabulary.

**Review by tomdeering-wf — APPROVED** [review]

**Review by michaelcarter-wf — APPROVED** [review]
+1

**Pull request merged**